presto-main/src/main/java/com/facebook/presto/operator/SpatialJoinOperator.java

```diff
@@ -120,7 +119,8 @@ public OperatorFactory duplicate()
     private final int probeGeometryChannel;
     private final PagesSpatialIndexFactory pagesSpatialIndexFactory;
 
-    private ListenableFuture<PagesSpatialIndex> pagesSpatialIndexFuture;
+    private ListenableFuture<Supplier<PagesSpatialIndex>> pagesSpatialIndexFuture;
```

Comment:
The factory can still return `ListenableFuture<PagesSpatialIndex>`, hiding the fact there are some `Supplier`s in play from the `SpatialJoinOperator` -- `SpatialJoinOperator` doesn't need to know about that.
The question is where you do `supplier.get()`. You can do it here, as you did, or you can do this in the factory, when satisfying the futures.

(for the record, in case of hash join the `supplier.get()` was in the factory -- before spill was implemented. Spill made things a bit different, so there isn't simple supplier anymore there.)